src/unistd.rs

```diff
@@ -1098,6 +1098,36 @@ pub fn close(fd: RawFd) -> Result<()> {
     Errno::result(res).map(drop)
 }
 
+#[cfg(target_os = "linux")]
+libc_bitflags! {
+    /// Options for close_range()
+    pub struct CloseRangeFlags : c_uint {
+        /// Unshare file descriptors before closing them.
+        CLOSE_RANGE_UNSHARE;
+        /// Set close-on-exec instead of closing file descriptors.
+        CLOSE_RANGE_CLOEXEC;
+    }
+}
+
+/// Close a range of raw file descriptors.
+/// [close_range(2)](https://man7.org/linux/man-pages/man2/close_range.2.html).
+#[cfg(target_os = "linux")]
+pub fn close_range(
```

Comment:
I think this function should be marked `unsafe` as it is kinda easy to get a double close in Rust, let's say you have created 3 `std::fs::File`s, which have file descriptors 3, 4, 5, then you want to close them with `close_range()`, to avoid a double free, your have to disable the close behavior of `std::fs::File` (or other things that are basically an `OwnedFd`), then the code would be something like this:

```rust
let file1: File1;
let file2: File2;
let file3: File3;

let fd1 = into_raw_fd(file1);
// if you forget to handle the file descriptors between [first, last]. then 
// double close would happen
let _  =  into_raw_fd(file2);
let fd3 = into_raw_fd(file3);

close_range(fd1, fd3, 0);
```
Due to its unsafety, it would be necessary to add a doc to remind the users, just like [`close()`'s document](https://docs.rs/nix/latest/nix/unistd/fn.close.html) does.

------

What about `close(2)`, why isn't it marked `unsafe`?

The current signature of `close()` is also not that safe:

```rust
fn close(fd: RawFd) -> Result<()>
```

And nix is working on [I/O safety](https://github.com/nix-rust/nix/issues/1750), in #1863, we will change its interface to:

```rust
fn close<Fd: IntoRawFd>(fd: Fd) -> Result<()>
```

`IntoRawFd::into_raw_fd(self)` would take `self` by value and `std::mem::forget()` the inner `OwnedFd`, which means that ONLY one `close()` would be called on that fd,  and is thus safe

Reply:
Yeah, unsafe makes sense here.  Especially because close_range is most useful to use post-fork and pre-exec, which is already a highly unsafe environment.

Reply:
@SteveLauC:
> I think this function should be marked unsafe as it is kinda easy to get a double close in Rust

I think we're all already on the same page, but just to say it explicitly: double-close is not a memory-safety issue (subsequent operations on the closed FDs would start returning EBADF), but it is an [I/O safety](https://github.com/rust-lang/rfcs/blob/master/text/3128-io-safety.md) issue.

@SteveLauC:
> The current signature of close() is also not that safe:
...
And nix is working on https://github.com/nix-rust/nix/issues/1750, in https://github.com/nix-rust/nix/pull/1863, we will change its interface to:
...

If this crate is working towards I/O safety then I agree this function cannot be made I/O safe and thus must be `unsafe`.

@asomers:
> Especially because close_range is most useful to use post-fork and pre-exec

My use case is actually not post-fork/pre-exec; I'm ensuring there are no unexpected (eg: inherited) FDs before enabling a SECCOMP filter and running some less-trusted code.

It's also possible to use this function with the `CLOSE_RANGE_CLOEXEC` flag in a way that doesn't affect I/O safety, but obviously the interface as-is doesn't restrict the caller to that (safe) use.

I've marked it `unsafe` in  3e5d88108d100b0c8f8483a4871d21d323eed1e